### PR TITLE
Introduce support for auto-magic service binding

### PR DESCRIPTION
**Pull request by iocanel**
This pull request introduces a zero configuration service binding solution for the services supported by the Service Binding Operator.

The idea is that we map each service kind to an Operators and its Custom Resource that is used to provision the service.

Each time we detect that a service is needed we generate a ServiceBidning that binds the `app` to the target `custom resource`. 
For unnamed services (e.g. unamed data sources) we use the service kind as name and id and the user is able to override it using configuration.

Remaining Tasks
- [x] Generate ServiceBinding per service (currently we generate one for all).
- [x] Add integration tests for mysql.
- [x] Add integration tests for semi-auto binding.

Future tasks
- Add kafka support (not possible atm).
- Review the default operators for each service type.

### CODE REVIEW

**Comment by iocanel** [comment]
@geoand: given that this PR touches kubernetes and kubernetes service binding a review would be welcome.
@maxandersen: as we have discussed about this on zulip in the past, it would be nice if you'd have a look.

**Comment by geoand** [comment]
Cool! I'll have a look tomorrow

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 1]
```diff
@@ -1312,6 +1312,96 @@ NOTE: Advanced, features (e.g. `ServiceBiniding` mappings) are not yet supported
...
+----
+
+For named datasources, the name of the datasource is used instead:
+
+[source,properties]
+----
+quarkus.datasource.fruits-db.db-kind=postgresql
+----
+
+The the `service` in the generated `ServiceBinding` will look like:
```
Comment (geoand) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1355]:
```suggestion
The `service` in the generated `ServiceBinding` will look like:
```

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 2]
```diff
@@ -1312,6 +1312,96 @@ NOTE: Advanced, features (e.g. `ServiceBiniding` mappings) are not yet supported
...
+    kind: PerconaXtraDBCluster
+    name: fruits-db
+----
+
+==== Customizing automatic bindings
+
+There are cases where customization upon the generated `ServiceBinding` may be needed. For example, the target resource name does not match the datasource name, or a non default operator needs to be used.
+For such cases, it's possible to use the `quarkus.kubernetes-service-binding.servies.<service id>.` properties and specify as service id the name of the service.
+
+The example below shows how to customize the `apiVerison` of the `PostgresCluster` resource:
```
Comment (geoand) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1387]:
```suggestion
The example below shows how to customize the `apiVersion` of the `PostgresCluster` resource:
```

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 3]
```diff
@@ -1312,6 +1312,96 @@ NOTE: Advanced, features (e.g. `ServiceBiniding` mappings) are not yet supported
...
+
+[source,properties]
+----
+quarkus.datasource.db-kind=postgresql
+quarkus.kubernetes-service-binding.services.postgresql.api-version: postgres-operator.crunchydata.com/v1beta2
```
Comment (geoand) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1392]: This looks wrong

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 4]
```diff
@@ -1312,6 +1312,96 @@ NOTE: Advanced, features (e.g. `ServiceBiniding` mappings) are not yet supported
...
+----
+
+Or in case that we have a named datasource:
+
+[source,properties]
+----
+quarkus.datasource.fruits-db.db-kind=postgresql
+quarkus.kubernetes-service-binding.services.fruits-db.api-version: postgres-operator.crunchydata.com/v1beta2
```
Comment (geoand) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1400]: Same here

**Review by geoand — COMMENTED** [review]
Added some doc related comments.

Also just to be absolutely clear: This does not add any runtime dependency, right?

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 4 above)
Comment (iocanel) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1400]: I am not sure I follow

**Comment by iocanel** [comment]
> Added some doc related comments.
> 
> Also just to be absolutely clear: This does not add any runtime dependency, right?

Right

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 4 above)
Comment (geoand) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1400]: This line doesn't look like a valid `properties` line - it looks like a mix between properties and YAML

**Review by geoand — APPROVED** [review]
LGTM, but let's also get @maxandersen's opinion

**Comment by maxandersen** [comment]
Why is it that Kafka is not possible atm ?

**Comment by iocanel** [comment]
> Why is it that Kafka is not possible atm ?

Based on this https://github.com/redhat-developer/service-binding-operator#known-bindable-operators kafka should only work via `Openshift Streams for Apache Kafka` but it's something that is not straight forward. 

So, there are two issues here.

1. Does it make sense to have as default resources for Kafka rhoas ones?
2. Does it make sense to have auto-magic support for something that is not straight forward anyway?

**Comment by iocanel** [comment]
@geoand @maxandersen: I finally have added the bits we discussed about (allowing the use of both `<service name>` and `<service kind>-<service name>` as configuration key. I think that we can merge.

**Comment by geoand** [comment]
👍🏼

**Comment by geoand** [comment]
I have already approved this one, so up to @maxandersen

**Comment by geoand** [comment]
@maxandersen since it would be great to have this in for `2.7.0.CR1`, I am going to merge it tonight unless you have any objections.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java [hunk 5]
```diff
@@ -193,7 +196,17 @@ private DeploymentResultBuildItem deploy(DeploymentTargetEntry deploymentTarget,
...
+                    r.createOrReplace();
+                } catch (Exception e) {
+                    if (e instanceof InterruptedException) {
+                        throw e;
+                    } else if (isOptional(optionalResourceDefinitions, i)) {
+                      log.warn("Failed to apply: " + i.getKind() + " " + i.getMetadata().getName() + ", possilby due to missing a CRD apiVersion: " + i.getApiVersion() + " and kind: " i.getKind() + ".");
+                    } else {
+                        throw e;
+                    }
+                }
```
Comment (geoand) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java:209]: Very nice!

**Pull request merged**